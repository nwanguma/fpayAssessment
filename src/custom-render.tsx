import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

const Wrapper: React.FC = ({ children }) => {
  return <MemoryRouter>{children}</MemoryRouter>;
};

const customRender = (ui: any, options?: any) =>
  render(ui, { wrapper: Wrapper, ...options });

export * from "@testing-library/react";

export { customRender as render };
