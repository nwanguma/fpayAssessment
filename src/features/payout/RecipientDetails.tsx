import { useState } from "react";

import validateFormData, { Error } from "../../validation/validateFormData";
import { Event, SubmitEvent, ToggleData } from "./payout";
import { State } from "./payout-reducer";

interface IProps {
  payoutDetails: State;
  handleOnChange: (e: Event) => void;
  handleProceed: () => void;
}

interface Classes {
  sectionHeading: string;
  tab: (region: number, id: number) => string;
  label: string;
  input: (error: boolean) => string;
  btn: string;
}

interface ValidationState {
  [key: string]: string;
}

const classes: Classes = {
  sectionHeading:
    "py-3 border-solid font-medium border-b border-white-border mb-5 text-sm leading-6 text-blue-text not-italic",
  tab: (region: number, id: number) =>
    `px-2 sm:px-6 text-[13px] sm:text-base cursor-pointer ${
      region === id
        ? " border-blue-tab font-medium  border-b-2 border-solid text-blue-tab"
        : "text-grey-label"
    } pb-3`,
  label: "label text-[13px] top-2",
  input: (error: boolean): string =>
    `${
      error ? "border-red-error" : "border-white-border-alt"
    } input h-[45px] py-3 px-[15px] text-grey-light`,
  btn: "btn w-[100%] text-white-main bg-blue-tab",
};

const Amount: React.FC<IProps> = ({
  payoutDetails,
  handleOnChange,
  handleProceed,
}) => {
  const [validationErrors, setValidationErrors] = useState<Error>({});
  const [recipientRegion, setRecipientRegion] = useState(1);

  const handleRegionChange = (id: number): void => {
    setRecipientRegion(id);
  };

  const handleSubmit = (e: SubmitEvent): void => {
    e.preventDefault();

    const { fullname, mail, accountNumber, swiftCode } = payoutDetails;

    const formState: ValidationState = {
      fullname,
      mail,
      accountNumber,
      swiftCode,
    };

    if (!swiftCode) delete formState.swiftCode;

    const keys = Object.keys(formState);
    const errors = validateFormData(formState, keys);

    setValidationErrors(errors);

    if (Object.keys(errors).length > 0) return;

    handleProceed();
  };

  return (
    <div className="card">
      <div className="container mb-5">
        <h3 className="card-heading__primary">Your Recipient</h3>
        <p className="card-heading__secondary">Who are you sending money to?</p>
      </div>
      <form onSubmit={handleSubmit} autoComplete="off">
        <div className="container relative h-[65px]">
          <label htmlFor="amount" className={classes.label}>
            Their email (optional)
          </label>
          <input
            name="mail"
            value={payoutDetails.mail}
            autoComplete="user-address"
            onChange={handleOnChange}
            type="text"
            className={classes.input(
              validationErrors.mail && validationErrors.mail.error
            )}
          ></input>
        </div>
        <div className="container relative h-[65px] mt-4">
          <label htmlFor="amount" className={classes.label}>
            Full name of the account holder
          </label>
          <input
            name="fullname"
            autoComplete="user-fullname"
            value={payoutDetails.fullname}
            onChange={handleOnChange}
            type="text"
            className={classes.input(
              validationErrors.fullname && validationErrors.fullname.error
            )}
          ></input>
        </div>
        <div className="container mt-5">
          <h3 className={classes.sectionHeading}>Bank Details</h3>
          <div className="flex">
            <p
              onClick={() => handleRegionChange(1)}
              className={classes.tab(recipientRegion, 1)}
            >
              Inside Europe
            </p>
            <p
              onClick={() => handleRegionChange(2)}
              className={classes.tab(recipientRegion, 2)}
            >
              Outside Europe
            </p>
          </div>
          {recipientRegion === 2 && (
            <div className="container relative h-[65px] mt-4">
              <label htmlFor="amount" className={classes.label}>
                SWIFT / BIC code
              </label>
              <input
                name="swiftCode"
                value={payoutDetails.swiftCode}
                onChange={handleOnChange}
                type="text"
                className={classes.input(
                  validationErrors.swiftCode && validationErrors.swiftCode.error
                )}
              ></input>
            </div>
          )}
          <div className="container relative h-[65px] mt-4">
            <label htmlFor="swiftCode" className={classes.label}>
              {recipientRegion === 1 ? "IBAN" : "IBAN / Account Number"}
            </label>
            <input
              name="accountNumber"
              value={payoutDetails.accountNumber}
              onChange={handleOnChange}
              type="text"
              className={classes.input(
                validationErrors.accountNumber &&
                  validationErrors.accountNumber.error
              )}
            ></input>
          </div>
        </div>
        <div className="container mt-8 flex justify-between">
          <button className={classes.btn}>Continue</button>
        </div>
      </form>
    </div>
  );
};

export default Amount;
