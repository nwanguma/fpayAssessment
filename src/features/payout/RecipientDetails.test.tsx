import { render, screen } from "../../custom-render";
import userEvent from "@testing-library/user-event";

import RecipientDetails from "./RecipientDetails";

describe("RecipientDetails", () => {
  const payoutDetails = {
    userAmount: "",
    feeRate: 0.00369,
    fee: 0,
    amountMinusFee: 0,
    baseCurrency: "EUR",
    rates: {},
    rate: 0,
    targetCurrency: "NGN",
    convertedAmount: "",
    fullname: "",
    mail: "",
    accountNumber: "",
    swiftCode: "",
  };

  const handleOnChange = jest.fn();
  const handleProceed = jest.fn();

  test("<RecipientDetails /> matches snapshot", () => {
    const component = render(
      <RecipientDetails
        payoutDetails={payoutDetails}
        handleOnChange={handleOnChange}
        handleProceed={handleProceed}
      />
    );

    expect(component.container).toMatchSnapshot();
  });

  test("It renders correctly", () => {
    render(
      <RecipientDetails
        payoutDetails={payoutDetails}
        handleOnChange={handleOnChange}
        handleProceed={handleProceed}
      />
    );

    expect(screen.getByTestId("card")).toBeInTheDocument();
  });

  test("It contains a button", () => {
    render(
      <RecipientDetails
        payoutDetails={payoutDetails}
        handleOnChange={handleOnChange}
        handleProceed={handleProceed}
      />
    );

    expect(
      screen.getByRole("button", { name: "Continue" })
    ).toBeInTheDocument();
  });
});

describe("Test form", () => {
  const payoutDetails = {
    userAmount: "",
    feeRate: 0.00369,
    fee: 0,
    amountMinusFee: 0,
    baseCurrency: "EUR",
    rates: {},
    rate: 0,
    targetCurrency: "NGN",
    convertedAmount: "",
    fullname: "",
    mail: "",
    accountNumber: "",
    swiftCode: "",
  };

  const handleOnChange = jest.fn();
  const handleProceed = jest.fn();

  test("shows all required input fields with empty values", () => {
    const { getByTestId } = render(
      <RecipientDetails
        payoutDetails={payoutDetails}
        handleOnChange={handleOnChange}
        handleProceed={handleProceed}
      />
    );

    const userMail = getByTestId("user-mail") as HTMLInputElement;
    const userFullname = getByTestId("user-full") as HTMLInputElement;
    const accountNumber = getByTestId("account-number") as HTMLInputElement;

    expect(userMail.value).toBe("");
    expect(userFullname.value).toBe("");
    expect(accountNumber.value).toBe("");
  });
});

describe("Simulate input interaction", () => {
  const payoutDetails = {
    userAmount: "",
    feeRate: 0.00369,
    fee: 0,
    amountMinusFee: 0,
    baseCurrency: "EUR",
    rates: {},
    rate: 0,
    targetCurrency: "NGN",
    convertedAmount: "",
    fullname: "",
    mail: "",
    accountNumber: "",
    swiftCode: "",
  };

  const handleOnChange = jest.fn();
  const handleProceed = jest.fn();

  test("It should update input value with sample text", () => {
    const { getByTestId } = render(
      <RecipientDetails
        payoutDetails={payoutDetails}
        handleOnChange={handleOnChange}
        handleProceed={handleProceed}
      />
    );

    const input = getByTestId("user-mail") as HTMLInputElement;

    userEvent.type(input, "user@fliqpay.com");

    expect(input).toHaveValue("user@fliqpay.com");
  });
});

export {};
