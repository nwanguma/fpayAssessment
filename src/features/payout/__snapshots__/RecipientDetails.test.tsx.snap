// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipientDetails <RecipientDetails /> matches snapshot 1`] = `
<div>
  <div
    class="card"
    data-testid="card"
  >
    <div
      class="container mb-5"
    >
      <h3
        class="card-heading__primary"
      >
        Your Recipient
      </h3>
      <p
        class="card-heading__secondary"
      >
        Who are you sending money to?
      </p>
    </div>
    <form
      autocomplete="off"
    >
      <div
        class="container relative h-[65px]"
      >
        <label
          class="label text-[13px] top-2"
          for="amount"
        >
          Their email (optional)
        </label>
        <input
          autocomplete="user-mail"
          class="border-white-border-alt input h-[45px] py-3 px-[15px] text-grey-light"
          data-testid="user-mail"
          name="mail"
          type="text"
          value=""
        />
      </div>
      <div
        class="container relative h-[65px] mt-4"
      >
        <label
          class="label text-[13px] top-2"
          for="amount"
        >
          Full name of the account holder
        </label>
        <input
          autocomplete="user-full"
          class="border-white-border-alt input h-[45px] py-3 px-[15px] text-grey-light"
          data-testid="user-full"
          name="fullname"
          type="text"
          value=""
        />
      </div>
      <div
        class="container mt-5"
      >
        <h3
          class="py-3 border-solid font-medium border-b border-white-border mb-5 text-sm leading-6 text-blue-text not-italic"
        >
          Bank Details
        </h3>
        <div
          class="flex"
        >
          <p
            class="px-2 sm:px-6 text-[13px] sm:text-base cursor-pointer  border-blue-tab font-medium  border-b-2 border-solid text-blue-tab pb-3"
          >
            Inside Europe
          </p>
          <p
            class="px-2 sm:px-6 text-[13px] sm:text-base cursor-pointer text-grey-label pb-3"
          >
            Outside Europe
          </p>
        </div>
        <div
          class="container relative h-[65px] mt-4"
        >
          <label
            class="label text-[13px] top-2"
            for="swiftCode"
          >
            IBAN
          </label>
          <input
            class="border-white-border-alt input h-[45px] py-3 px-[15px] text-grey-light"
            data-testid="account-number"
            name="accountNumber"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="container mt-8 flex justify-between"
      >
        <button
          class="btn w-[100%] text-white-main bg-blue-tab"
        >
          Continue
        </button>
      </div>
    </form>
  </div>
</div>
`;
